### PR TITLE
Fix internal spec links

### LINKED ISSUE #37: Fix links
links didn't survive the move into Github. Fix them so that they lead the user to the appropriate sections.

### PR DESCRIPTION
**Pull request by lavaleri**
*Issue #, if available:* Closes #37 

*Description of changes:* Fixes the markdown links so that the spec can now correctly link to other parts of the spec.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: client-apis/decrypt.md [hunk 1]
```diff
@@ -79,42 +79,42 @@ The decrypted data.
...
 
-This behavior MUST obtain this set of [decryption materials](#structures.md#decryption-materials), 
-by calling [Decrypt Materials](#cmm-interface.md#decrypt-materials) on a [CMM](#cmm-interface.md).
+This behavior MUST obtain this set of [decryption materials](../framework/structures.md#decryption-materials), 
+by calling [Decrypt Materials](../framework/cmm-interface.md#decrypt-materials) on a [CMM](../framework/cmm-interface.md).
 
 The CMM used MUST be the input CMM, if supplied.
-If a CMM is not supplied as the input, the decrypt behavior MUST construct a [default CMM](#default-cmm.md) 
+If a CMM is not supplied as the input, the decrypt behavior MUST construct a [default CMM](../framework/default-cmm.md) 
 from the [keyring](#keyring) inputted.
```
Comment (mattsb42-aws) [inline client-apis/decrypt.md]:
```suggestion
from the [keyring](../framework/keyring-interface.md) inputted.
```

File: client-apis/decrypt.md [hunk 2]
```diff
@@ -79,42 +79,42 @@ The decrypted data.
...
 
-The call to CMM's [Decrypt Materials](#cmm-interface.md#decrypt-materials) behavior MUST include as the input the 
-[encryption context](#message-header.md#aad), if provided, the [encrypted data keys](#message-header.md#encrypted-data-keys) and the 
-[algorithm suite ID](#message-header.md#algorithm-suites-id), obtained from parsing the message header of the encrypted message inputted.
+The call to CMM's [Decrypt Materials](../framework/cmm-interface.md#decrypt-materials) behavior MUST include as the input the 
+[encryption context](../data-format/message-header.md#aad), if provided, the [encrypted data keys](../data-format/message-header.md#encrypted-data-keys) and the 
+[algorithm suite ID](../data-format/message-header.md#algorithm-suites-id), obtained from parsing the message header of the encrypted message inputted.
```
Comment (mattsb42-aws) [inline client-apis/decrypt.md]:
```suggestion
[algorithm suite ID](../data-format/message-header.md#algorithm-suite-id), obtained from parsing the message header of the encrypted message inputted.
```

File: client-apis/decrypt.md [hunk 3]
```diff
@@ -79,42 +79,42 @@ The decrypted data.
...
 
 The decryption materials returned by the call to the CMM's Decrypt Materials behaviour MUST contain a valid 
-[plaintext data key](#structures.md#decryption-materials#plaintext-data-key),
-[algorithm suite](#structures.md#decryption-materials#algorithm-suite) and an
-[encryption context](#structures.md#decryption-materials#encryption-context), if an encryption context was used during encryption.    
+[plaintext data key](../framework/structures.md#decryption-materials#plaintext-data-key),
+[algorithm suite](../framework/structures.md#decryption-materials#algorithm-suite) and an
+[encryption context](../framework/structures.md#decryption-materials#encryption-context), if an encryption context was used during encryption.
```
Comment (mattsb42-aws) [inline client-apis/decrypt.md]:
```suggestion
[plaintext data key](../framework/structures.md#plaintext-data-key),
[algorithm suite](../framework/algorithm-suites.md) and an
[encryption context](../framework/structures.md#encryption-context), if an encryption context was used during encryption.    
```

File: client-apis/decrypt.md [hunk 4]
```diff
@@ -79,42 +79,42 @@ The decrypted data.
...
+[plaintext data key](../framework/structures.md#decryption-materials#plaintext-data-key),
+[algorithm suite](../framework/structures.md#decryption-materials#algorithm-suite) and an
+[encryption context](../framework/structures.md#decryption-materials#encryption-context), if an encryption context was used during encryption.    
 Note: This encryption context MUST be the same encryption context that was used during encryption otherwise the decrypt operation will fail.   
 
 The decrypt behavior MUST then use this plaintext data key, algorithm suite and encryption context, if included, to decrypt the encrypted content 
-and obtain the plaintext to be returned. The [encrypted content](#message-body.md#encrypted-content) to be decrypted is obtained by parsing the 
-[message body](#message-crypto.md) of the encrypted message inputted.   
+and obtain the plaintext to be returned. The [encrypted content](../data-format/message-body.md#encrypted-content) to be decrypted is obtained by parsing the 
+[message body](../framework/message-body.md) of the encrypted message inputted.
```
Comment (mattsb42-aws) [inline client-apis/decrypt.md]:
```suggestion
[message body](../data-format/message-body.md) of the encrypted message inputted.   
```

File: client-apis/encrypt.md [hunk 5]
```diff
@@ -41,21 +41,21 @@ The plaintext to encrypt.
 
 ### Encryption Context
 
-See [encryption context](#structures.md#encryption-context).
+See [encryption context](../framework/structures.md#encryption-context).
 
-The encryption context MUST NOT contain an entry with a [reserved key value](#data-structutes.md#encryption-context).
+The encryption context MUST NOT contain an entry with a [reserved key value](data-structutes.md#encryption-context).
```
Comment (mattsb42-aws) [inline client-apis/encrypt.md]:
```suggestion
The encryption context MUST NOT contain an entry with a [reserved key value](../framework/structures.md#encryption-context).
```

File: client-apis/encrypt.md [hunk 6]
```diff
@@ -66,89 +66,89 @@ A bound on the length of the [plaintext](#plaintext) to encrypt.
...
 constructed using the user supplied keyring as input.
-The call to [Get Encryption Materials](#cmm-interface.md#get-encryption-materials) MUST include the
+The call to [Get Encryption Materials](../framework/cmm-interface.md#get-encryption-materials) MUST include the
 input [encryption context](#encryption-context), if supplied.
 If the length is known on the input [plaintext](#plaintext), this call MUST also include that value
 
-The [algorithm suite](#algorithm-suites.md) used in all aspects of this behavior MUST be the algorithm suite in the
-[encryption materials](#encryption-materials.md) returned from the [Get Encryption Materials](#cmm-interface#get-encryption-materials) call.
+The [algorithm suite](../framework/algorithm-suites.md) used in all aspects of this behavior MUST be the algorithm suite in the
+[encryption materials](../framework/structures#encryption-materials.md) returned from the [Get Encryption Materials](../framework/cmm-interface.md#get-encryption-materials) call.
```
Comment (mattsb42-aws) [inline client-apis/encrypt.md]:
```suggestion
[encryption materials](../framework/structures.md#encryption-materials) returned from the [Get Encryption Materials](../framework/cmm-interface.md#get-encryption-materials) call.
```

File: client-apis/encrypt.md [hunk 7]
```diff
@@ -66,89 +66,89 @@ A bound on the length of the [plaintext](#plaintext) to encrypt.
...
-- [Encrypted Data Keys](#message-header.md#encrypted-data-key): MUST be the serialization of the
-  [encrypted data keys](#structures.md#encrypted-data-keys) in the [encryption materials](#structures.md#encryption-materials)
-- [Content Type](#message-header.md#content-type): MUST be `Framed`
-- [IV Length](#message-header.md#iv-length): MUST match the [IV length](#algorithm-suites.md#iv-length)
-  specified by the [algorithm suite](#algorithm-suites.md)
-- [Frame Length](#message-header.md#frame-length): MUST be the same value as the input [frame length](#frame-length),
+- [AAD](../data-format/message-header.md#aad): MUST be the serialization of the [encryption context](../framework/structures.md#encryption-context)
+  in the [encryption materials](../framework/structures.md#encryption-materials)
+- [Encrypted Data Keys](../data-format/message-header.md#encrypted-data-key): MUST be the serialization of the
+  [encrypted data keys](../framework/structures.md#encrypted-data-keys) in the [encryption materials](../framework/structures.md#encryption-materials)
```
Comment (mattsb42-aws) [inline client-apis/encrypt.md]:
```suggestion
- [Encrypted Data Keys](../data-format/message-header.md#encrypted-data-key-entries): MUST be the serialization of the
  [encrypted data keys](../framework/structures.md#encrypted-data-keys) in the [encryption materials](../framework/structures.md#encryption-materials)
```

File: client-apis/encrypt.md [hunk 8]
```diff
@@ -66,89 +66,89 @@ A bound on the length of the [plaintext](#plaintext) to encrypt.
...
+- [Content Type](../data-format/message-header.md#content-type): MUST be `Framed`
```
Comment (mattsb42-aws) [inline client-apis/encrypt.md]: For consistency:

```suggestion
- [Content Type](../data-format/message-header.md#content-type): MUST be [01](../data-format/message-header.md#supported-content-types)
```

File: client-apis/encrypt.md [hunk 9]
```diff
@@ -66,89 +66,89 @@ A bound on the length of the [plaintext](#plaintext) to encrypt.
...
+the output [message](../data-format/message.md) MUST contain a [message footer](../data-format/message-footer.md).
 The footer is serialized with the following specifics:
 
-- [Signature](#message-footer.md#signature): MUST be the output of the [signature algorithm](#algorithm-suites.md#signature-algorithm)
-  specified by the [algorithm suite](#algorithm-suites.md), with the following input:
-  - the signature key is the [signing key](#structures.md#signing-key) in the [encryption materials](#structures.md#encryption-materials)
+- [Signature](../data-format/message-footer.md#signature): MUST be the output of the [signature algorithm](../framework/algorithm-suites.md#signature-algorithm)
+  specified by the [algorithm suite](../framework/algorithm-suites.md), with the following input:
+  - the signature key is the [signing key](../framework/structures.md#signing-key) in the [encryption materials](../framework/structures.md#encryption-materials)
   - the input to sign is the concatenation of the serialization of the [message header](#message-header) and [message body](#message-body)
```
Comment (mattsb42-aws) [inline client-apis/encrypt.md]:
```suggestion
  - the input to sign is the concatenation of the serialization of the [message header](../data-format/message-header.md) and [message body](../data-format/message-body.md)
```

File: client-apis/encrypt.md [hunk 10]
```diff
@@ -66,89 +66,89 @@ A bound on the length of the [plaintext](#plaintext) to encrypt.
...
 
-Any data that is not specified within the [message format](#message.md) MUST NOT be added to the output message.
+Any data that is not specified within the [message format](../data-format/message.md) MUST NOT be added to the output message.
 
 ### Keyring Trace
 
-The [keyring trace](#structures.md#keyring-trace) obtained from the [encryption materials](#data-structure.md#encryption-materials).
+The [keyring trace](../framework/structures.md#keyring-trace) obtained from the [encryption materials](data-structure.md#encryption-materials).
```
Comment (mattsb42-aws) [inline client-apis/encrypt.md]:
```suggestion
The [keyring trace](../framework/structures.md#keyring-trace) obtained from the [encryption materials](../framework/structures.md#encryption-materials).
```

File: client-apis/encrypt.md [hunk 11]
```diff
@@ -166,15 +166,15 @@ If size is not known, how do we set the bounds?
 Implementations of the AWS Encryption SDK MUST not encrypt using the Non-Framed content type.
```
Comment (mattsb42-aws) [inline client-apis/encrypt.md]:
```suggestion
Implementations of the AWS Encryption SDK MUST NOT encrypt using the Non-Framed content type.
```

File: client-apis/encrypt.md [hunk 12]
```diff
@@ -166,15 +166,15 @@ If size is not known, how do we set the bounds?
...
-  - The AAD is the serialized [message body AAD](#message-body-aad.md)
-  - The IV is the [IV](#message-body.md#iv) specified above.
+- [IV](../data-format/message-body.md#iv):  MUST be the [sequence number](../data-format/message-body-aad.md#sequence-number)
+  used in the [message body AAD](../data-format/message-body-aad.md).
+- [Encrypted Content](../data-format/message-body.md#encrypted-content): MUST be the output of the [authenticated encryption algorithm](../framework/algorithm-suites.md#encryption-algorithm)
+  specified by the [algorithm suite](../framework/algorithm-suites.md), with the following inputs:
+  - The AAD is the serialized [message body AAD](../data-format/message-body-aad.md)
+  - The IV is the [IV](../data-format/message-body.md#iv) specified above.
   - The cipherkey is the derived data key
   - The plaintext is the input [plaintext message](#plaintext-message)
```
Comment (mattsb42-aws) [inline client-apis/encrypt.md]:
```suggestion
  - The plaintext is the input [plaintext](#plaintext)
```

File: data-format/message-body-aad.md [hunk 13]
```diff
@@ -32,23 +32,23 @@ The bytes are appended in the order shown.
...
 
 ### Sequence Number
 
 The sequence number of the data this message body AAD belongs to.
-For [framed data](#message-body.md#framed-data), the value of this field MUST be the [frame sequence number](#message-body.md#frame-sequence-number).
-For [non-framed data](#message-body.md#framed-data), the value of this field MUST be `1`.
+For [framed data](message-body.md#framed-data), the value of this field MUST be the [frame sequence number](message-body.md#sequence-number).
+For [non-framed data](message-body.md#non-framed-data), the value of this field MUST be `1`.
 
 ### Content Length
```
Comment (mattsb42-aws) [inline data-format/message-body-aad.md]: The frame length links below (out of comment range) are broken.

File: data-format/message-body.md [hunk 14]
```diff
@@ -16,9 +16,9 @@ See [Message Version](#message.md#version).
...
 ## Overview
 
-The message body is a component of the [message](#message.md).    
-The message body contains the encrypted data, called the [encrypted content](#encrypted-content.md).    
-The purpose of the message body is to define the structure containing the [encrypted content](#encrypted-content.md).
+The message body is a component of the [message](message.md).    
+The message body contains the encrypted data, called the [encrypted content](#encrypted-content).    
+The purpose of the message body is to define the structure containing the [encrypted content](#encrypted-content).
 
 The structure of the body depends on the content type:
```
Comment (mattsb42-aws) [inline data-format/message-body.md]: Juuust out of range of the comment space again, but the links for framed and non-framed data are swapped.

File: data-format/message-body.md [hunk 15]
```diff
@@ -60,11 +60,11 @@ The IV MUST be a unique IV within the message.
 
 The length of the encrypted content.    
 The length MUST NOT be greater than `2^36 - 32`, or 64 gibibytes (64 GiB), 
-due to restrictions imposed by the [implemented algorithms](#algorithm-suite.md).
+due to restrictions imposed by the [implemented algorithms](../framework/algorithm-suites.md).
 
 #### Encrypted Content
 
-The encrypted data as returned by the [encryption algorithm](#encrypt.md).
+The encrypted data as returned by the [encryption algorithm](../client-apis/encrypt.md).
```
Comment (mattsb42-aws) [inline data-format/message-body.md]:
```suggestion
The encrypted data as returned by the [encryption algorithm](../framework/algorithm-suites.md#encryption-algorithm).
```

File: data-format/message-body.md [hunk 16]
```diff
@@ -109,28 +109,28 @@ Subsequent frames MUST be in order and MUST contain an increment of 1 from the p
...
-The IV length MUST be equal to the IV length of the algorithm suite specified by the [Algorithm Suite ID](#message-header.md#algorithm-suite-id) field.
-Note: This IV is different from the [Header IV](#message-header.md#iv).  
+Each frame in the [Framed Data](#framed-data) MUST include an IV that is unique within the message.
+The IV length MUST be equal to the IV length of the algorithm suite specified by the [Algorithm Suite ID](message-header.md#algorithm-suite-id) field.
+Note: This IV is different from the [Header IV](message-header.md#iv).  
 
 ##### Encrypted Content
 
-The encrypted data for each frame, as returned by the [encryption algorithm](#algorithm-suite.md).  
+The encrypted data for each frame, as returned by the [encryption algorithm](../framework/algorithm-suites.md).
```
Comment (mattsb42-aws) [inline data-format/message-body.md]:
```suggestion
The encrypted data for each frame, as returned by the [encryption algorithm](../framework/algorithm-suites.md#encryption-algorithm).  
```

File: data-format/message-body.md [hunk 17]
```diff
@@ -147,14 +147,14 @@ For example, in the case that the length of the Plaintext is equal to the Frame
...
+Field                                          | Length (bytes)                                                                                               | Interpreted as
+---------------------------------------------- | ------------------------------------------------------------------------------------------------------------ | --------------
+[Sequence Number End](#sequence-number-end)    | 4                                                                                                            | Bytes
+[Sequence Number](#sequence-number)            | 4                                                                                                            | UInt32
+[IV](#iv)                                      | [IV Length](message-header.md#iv-length)                                                                     | Bytes
+[Encrypted Content Length](#encrypted-content) | 4                                                                                                            | UInt32
```
Comment (mattsb42-aws) [inline data-format/message-body.md]:
```suggestion
[Encrypted Content Length](#encrypted-content-length-1) | 4                                                                                                            | UInt32
```

File: data-format/message-body.md [hunk 18]
```diff
@@ -173,23 +173,23 @@ The Final Frame Sequence number MUST be equal to the total number of frames in t
 
 The initialization vector for the final frame.    
 The IV MUST be a unique IV within the message.   
-The IV length MUST be equal to the IV length of the [algorithm](#algorithm-suite.md) that generated the message. 
-Note: This IV is different from the [Header IV](#message-header.md#iv). 
+The IV length MUST be equal to the IV length of the [algorithm](../framework/algorithm-suites.md) that generated the message.
```
Comment (mattsb42-aws) [inline data-format/message-body.md]:
```suggestion
The IV length MUST be equal to the IV length of the [algorithm suite](../framework/algorithm-suites.md) that generated the message. 
```

File: data-format/message-body.md [hunk 19]
```diff
@@ -173,23 +173,23 @@ The Final Frame Sequence number MUST be equal to the total number of frames in t
...
+Note: This IV is different from the [Header IV](message-header.md#iv). 
 
 ##### Encrypted Content Length
 
 The length of the encrypted content.  
 
 ##### Encrypted Content
 
-The encrypted data for the final frame, as returned by the [encryption algorithm](#algorithm-suite.md).  
+The encrypted data for the final frame, as returned by the [encryption algorithm](../framework/algorithm-suites.md).
```
Comment (mattsb42-aws) [inline data-format/message-body.md]:
```suggestion
The encrypted data for the final frame, as returned by the [encryption algorithm](../framework/algorithm-suites.md#encryption-algorithm).
```

File: data-format/message-header.md [hunk 20]
```diff
@@ -16,13 +16,13 @@ See [Message Version](#message.md#version).
 
 ## Overview
 
-The message header is a component of the [message](#message.md).
+The message header is a component of the [message](message.md).
 
 The purpose of the message header is to define the authenticated metadata required for decryption of the [message body](#message-body), including:
```
Comment (mattsb42-aws) [inline data-format/message-header.md]:
```suggestion
The purpose of the message header is to define the authenticated metadata required for decryption of the [message body](message-body.md#encrypted-content), including:
```

File: data-format/message-header.md [hunk 21]
```diff
@@ -16,13 +16,13 @@ See [Message Version](#message.md#version).
...
 
-- the format of the [message](#message.md) and [message body](#message-body.md)
-- the encrypted data keys needed for decryption of the [message body encrypted content](#message-body.md#encrypted-content)
-- the [encryption context](#encryption-context.md)
+- the format of the [message](message.md) and [message body](message-body.md)
+- the encrypted data keys needed for decryption of the [message body encrypted content](message-body.md#encrypted-content)
+- the [encryption context](../framework/structures.md#encryption-context)
 
 ## Definitions
```
Comment (mattsb42-aws) [inline data-format/message-header.md]: Out of range again, but the "supported algorithm suites" link in the table should point to `../framework/algorithm-suites.md#supported-algorithm-suites`.

File: data-format/message-header.md [hunk 22]
```diff
@@ -100,13 +100,13 @@ The purpose of the message ID is to:
 - protect against accidental reuse of a derived data key or the wearing out of derived data keys in
   the AWS Encryption SDK
 - uniquely identify the [message](message.md)
-- bind the message header to the [message body](message-body.md) [TODO "bind" wording]
+- bind the message header to the [message body](#dtat-format/message-body.md) [TODO "bind" wording]
```
Comment (mattsb42-aws) [inline data-format/message-header.md]:
```suggestion
- bind the message header to the [message body](message-body.md) [TODO "bind" wording]
```

File: data-format/message-header.md [hunk 23]
```diff
@@ -120,12 +120,12 @@ The bytes are appended in the order shown.
 
 The length of the [Key Value Pairs](#key-value-pairs) field in bytes.
 
-When the [encryption context](#encryption-context.md) is empty, the value of this field MUST be 0.
+When the [encryption context](../framework/structures.md#encryption-context) is empty, the value of this field MUST be 0.
 
 ##### Key Value Pairs
 
 The encoding of the key-value pairs of the [encryption context](#encryption-context).
```
Comment (mattsb42-aws) [inline data-format/message-header.md]:
```suggestion
The encoding of the key-value pairs of the [encryption context](../framework/structures.md#encryption-context).
```

File: client-apis/encrypt.md [hunk 24]
```diff
@@ -66,89 +66,89 @@ A bound on the length of the [plaintext](#plaintext) to encrypt.
...
-  [encrypted data keys](#structures.md#encrypted-data-keys) in the [encryption materials](#structures.md#encryption-materials)
-- [Content Type](#message-header.md#content-type): MUST be `Framed`
-- [IV Length](#message-header.md#iv-length): MUST match the [IV length](#algorithm-suites.md#iv-length)
-  specified by the [algorithm suite](#algorithm-suites.md)
-- [Frame Length](#message-header.md#frame-length): MUST be the same value as the input [frame length](#frame-length),
+- [AAD](../data-format/message-header.md#aad): MUST be the serialization of the [encryption context](../framework/structures.md#encryption-context)
+  in the [encryption materials](../framework/structures.md#encryption-materials)
+- [Encrypted Data Keys](../data-format/message-header.md#encrypted-data-key-entries): MUST be the serialization of the
+  [encrypted data keys](../framework/structures.md#encrypted-data-keys) in the [encryption materials](../framework/structures.md#encryption-materials)
+- [Content Type](../data-format/message-header.md#content-type): MUST be [01](../data-format/message-header.md#supported-content-types)
```
Comment (mattsb42-aws) [inline client-apis/encrypt.md]: Sorry, that was my fault: misremembered values.
```suggestion
- [Content Type](../data-format/message-header.md#content-type): MUST be [02](../data-format/message-header.md#supported-content-types)
```

File: data-format/message-header.md [hunk 25]
```diff
@@ -255,21 +255,21 @@ The header authentication contains fields used for authentication of the [header
...
-The [initialization vector (IV)](#initialization-vector.md) used as input to calculate the [authentication tag](#authentication-tag).
+The initialization vector (IV) used as input to calculate the [authentication tag](#authentication-tag).
 
 #### Authentication Tag
 
 The authentication value for the header.
-The [algorithm suite](#algorithm-suites-id) specified by the [Algorithm Suite ID](#algorith-suite-id) field
-[determines how the value of this field is calculated](#encrypt.md),
-and uses this value to [authenticate the contents of the header during decryption](#decrypt.md).
+The [algorithm suite](../framework/algorithm-suites.md) specified by the [Algorithm Suite ID](#algorith-suite-id) field
```
Comment (mattsb42-aws) [inline data-format/message-header.md]:
```suggestion
The [algorithm suite](../framework/algorithm-suites.md) specified by the [Algorithm Suite ID](#algorithm-suite-id) field
```

File: framework/multi-keyring.md [hunk 26]
```diff
@@ -60,78 +60,78 @@ If this keyring does not have a [generator keyring](#generator-keyring), this li
 ### OnEncrypt
 
 If this keyring has a [generator keyring](#generator-keyring),
-this keyring MUST first call that generator keyring's [OnEncrypt](#keyring-interface#onencrypt.md)
-using the input [encryption materials](#structures.md#encryption-materials) as input.
-If the [generator keyring](#generator-keyring) fails [OnEncrypt](#keyring-interface.md), this OnEncrypt MUST also fail.
+this keyring MUST first call that generator keyring's [OnEncrypt](keyring-interface.md#onencrypt)
+using the input [encryption materials](structures.md#encryption-materials) as input.
+If the [generator keyring](#generator-keyring) fails [OnEncrypt](keyring-interface.md#onencrypt), this OnEncrypt MUST also fail.
 If the [generator keyring](#generator-keyring) returns [encryption materials](#encryption-materials) missing a plaintext data key,
```
Comment (mattsb42-aws) [inline framework/multi-keyring.md]:
```suggestion
If the [generator keyring](#generator-keyring) returns [encryption materials](structures.md#encryption-materials) missing a plaintext data key,
```

File: framework/raw-aes-keyring.md [hunk 27]
```diff
@@ -124,59 +124,59 @@ The authentication tag outputted from the AES-GCM encryption.
...
+The keyring MUST encrypt the plaintext data key in the [encryption materials](structures.md#encryption-materials)
 using AES-GCM.
 
 The keyring must use AES-GCM with the following specifics:
 
-- it uses the [encryption context in the encryption materials](#structures.md#encryption-materials)
+- it uses the [encryption context](structures.md#encryption-context-1) in the encryption materials
   as the additional authenticated data (AAD)
-- the AAD is serialized in the same format as the serialization of [message header AAD key value pairs](#message-header.md#key-value-pairs)
+- the AAD is serialized in the same format as the serialization of [message header AAD key value pairs](#data-format/message-header.md#key-value-pairs)
```
Comment (mattsb42-aws) [inline framework/raw-aes-keyring.md]:
```suggestion
- the AAD is serialized in the same format as the serialization of [message header AAD key value pairs](../data-format/message-header.md#key-value-pairs)
```

File: framework/raw-aes-keyring.md [hunk 28]
```diff
@@ -192,20 +192,20 @@ The keyring MUST attempt to decrypt the encrypted data key if and only if the fo
...
 
-- it uses the [encrypted-key](#encrypted-key) obtained from deserialization as the AES-GCM input ciphertext.
+- it uses the [encrypt key](#encrypted-key) obtained from deserialization as the AES-GCM input ciphertext.
 - it uses the [authentication tag](#authentication-tag) obtained from deserialization as the AES-GCM input authentication tag.
 - it uses this keyring's [wrapping key](#wrapping-key) as the AES-GCM cipher key.
 - it uses the [IV](#iv) obtained from deserialization as the AES-GCM IV.
-- it uses the encryption context from the [decryption materials](#structures.md#decryption-materials) as the AES-GCM AAD.
-- the AAD is serialized in the same format as the serialization of [message header AAD key value pairs](#message-header.md#key-value-pairs)
+- it uses the encryption context from the [decryption materials](structures.md#decryption-materials) as the AES-GCM AAD.
+- the AAD is serialized in the same format as the serialization of [message header AAD key value pairs](#data-format/message-header.md#key-value-pairs)
```
Comment (mattsb42-aws) [inline framework/raw-aes-keyring.md]:
```suggestion
- the AAD is serialized in the same format as the serialization of [message header AAD key value pairs](../data-format/message-header.md#key-value-pairs)
```

File: framework/raw-rsa-keyring.md [hunk 29]
```diff
@@ -91,68 +91,68 @@ The private key SHOULD contain all Chinese Remainder Theorem (CRT) components (p
```
Comment (mattsb42-aws) [inline framework/raw-rsa-keyring.md]: Out of range:

The Browser JS link is broken. Should be `https://github.com/aws/aws-encryption-sdk-javascript/blob/master/modules/raw-rsa-keyring-browser/src/raw_rsa_keyring_web_crypto.ts`.

File: framework/raw-rsa-keyring.md [hunk 30]
```diff
@@ -91,68 +91,68 @@ The private key SHOULD contain all Chinese Remainder Theorem (CRT) components (p
...
+If the [encryption materials](structures.md#encryption-materials) do not contain a plaintext data key,
+on encrypt MUST generate a random plaintext data key and set it on the [encryption materials](structures.md#encryption-materials).
 
 The keyring MUST attempt to encrypt the plaintext data key in the
-[encryption materials](#structures.md#encryption-materials) using RSA.
+[encryption materials](structures.md#encryption-materials) using RSA.
 The keyring performs RSA with the following specifics:
 
 - this keyring's [public key](#public-key) is the RSA public key
 - this keyring's [padding scheme](padding-scheme) is the RSA padding scheme.
```
Comment (mattsb42-aws) [inline framework/raw-rsa-keyring.md]:
```suggestion
- this keyring's [padding scheme](#supported-padding-schemes) is the RSA padding scheme.
```

File: framework/raw-rsa-keyring.md [hunk 31]
```diff
@@ -91,68 +91,68 @@ The private key SHOULD contain all Chinese Remainder Theorem (CRT) components (p
...
 
 If RSA encryption was successful, OnEncrypt MUST return the input
-[encryption materials](#structures.md#encryption-materials), modified in the following ways:
+[encryption materials](structures.md#encryption-materials), modified in the following ways:
 
 - The encrypted data key list has a new encrypted data key added, constructed as follows:
-  - the [key provider ID](#structures.md#key-provider-id) field is this keyring's [key namespace](#key-id).
-  - the [key provider information](#structures.md#key-provider-information) field is this keyring's [key name](#key-name).
-  - the [ciphertext](#structures.md#data-key-encryption) field is the ciphertext outputted from
+  - the [key provider ID](structures.md#key-provider-id) field is this keyring's [key namespace](#key-id).
```
Comment (mattsb42-aws) [inline framework/raw-rsa-keyring.md]:
```suggestion
  - the [key provider ID](structures.md#key-provider-id) field is this keyring's [key namespace](#key-namespace).
```

File: framework/raw-rsa-keyring.md [hunk 32]
```diff
@@ -91,68 +91,68 @@ The private key SHOULD contain all Chinese Remainder Theorem (CRT) components (p
...
+- the encrypted data key's [key provider information](structures.md#key-provider-information)
   has a value equal to this keyring's [key name](#key-name).
-- the encrypted data key's [key provider ID](#structures.md#key-provider-id) has a value equal to
+- the encrypted data key's [key provider ID](structures.md#key-provider-id) has a value equal to
   this keyring's [key namespace](#key-namespace).
 
 The keyring performs RSA decryption with the following specifics:
 
 - this keyring's [private key](#private-key) is the RSA private key
 - this keyring's [padding scheme](padding-scheme) is the RSA padding scheme.
```
Comment (mattsb42-aws) [inline framework/raw-rsa-keyring.md]:
```suggestion
- this keyring's [padding scheme](#supported-padding-schemes) is the RSA padding scheme.
```

File: framework/structures.md [hunk 33]
```diff
@@ -196,35 +196,35 @@ The value of this key MUST be kept secret.
...
 
-Decryption materials are a structure containing materials needed for [decryption](#decrypt.md).
+Decryption materials are a structure containing materials needed for [decryption](../client-apis/decrypt.md).
 This structure MAY include any of the following fields:
 
-- [Algorithm Suite](#algorith-suite)
-- [Encryption Context](#encryption-context)
-- [Keyring Trace](#keyring-trace)
-- [Plaintext Data Key](#plaintext-data-key)
+- [Algorithm Suite](#algorith-suite-1)
```
Comment (mattsb42-aws) [inline framework/structures.md]:
```suggestion
- [Algorithm Suite](#algorithm-suite-1)
```

File: framework/caching-cmm.md [hunk 34]
```diff
@@ -102,48 +102,48 @@ the caching CMM MUST update the usage stats for the cache entry retrieved.
...
 from the underlying [cryptographic materials cache (CMC)](#underlying-cryptographic-materials-cache).  
 
 If a cache entry is found, the caching CMM MUST return the encryption materials retrieved.
 If a cache entry is not found, the caching CMM MUST then attempt to obtain the encryption materials 
-by making a call to the underlying CMM's [Get Encryption Materials](#cmm-interface.md#get-encryption-materials).  
+by making a call to the underlying CMM's [Get Encryption Materials](cmm-interface.md#get-encryption-materials).  
 
-If the [algorithm suite](#algorithm-suites.md) requested contains a [Identity KDF](#algorithm-suites.md#identity-kdf),
+If the [algorithm suite](algorithm-suites.md) requested contains a [Identity KDF](algorithm-suites.md#identity-kdf),
 the caching CMM MUST add the encryption materials obtained from the underlying CMM into the underlying CMC.
```
[the remark this answers is not in the document]
Comment (lavaleri) [reply framework/caching-cmm.md]: I'll break into a new PR. Thanks for finding!

File: framework/cryptographic-materials-cache.md [hunk 35]
```diff
@@ -38,7 +38,7 @@ The cache identifier used to uniquely identify a single cache entry in the crypt
 ### Cache Entry 
 
 A cache entry represents an entry in the cryptographic materials cache.  
-A cache entry contains the cryptographic materials, [encryption](#structures.md#encryption-materials) or [decryption](#structures.md#decryption-materials) 
+A cache entry contains the cryptographic materials, [encryption](structures.md#encryption-materials) or [decryption](structures.md#decryption-materials) 
 materials to be cached along with other metadata.  
 Some implementations of the CMC have the following metadata:
```
[the remark this answers is not in the document]
Comment (lavaleri) [reply framework/cryptographic-materials-cache.md]: > The links to encryption materials under "Usage Metadata" are broken.

Fixed.



> The usage metadata links are broken. NOTE: This is affected by the change from "Message Encrypted" to "Messages Encrypted".

I'm not sure what you mean by this

[the remark this answers is not in the document]
Comment (mattsb42-aws) [reply framework/cryptographic-materials-cache.md]: NVM, I was reading the metadata values as referencing messages/bytes encrypted values in the caching CMM spec, not separate values in this spec. Nothing else to fix here.

File: framework/caching-cmm.md (hunk 34 above)
[the remark this answers is not in the document]
Comment (lavaleri) [reply framework/caching-cmm.md]: https://github.com/awslabs/aws-encryption-sdk-specification/pull/88

**Pull request merged**